tfhe/src/integer/server_key/radix_parallel/add.rs

```diff
@@ -139,8 +153,335 @@ impl ServerKey {
+
+        let num_blocks = carry_out.len();
+        let num_steps = (carry_out.len() as f64).log2() as usize;
+
+        let lut_carry_propagation_sum = self.key.generate_accumulator_bivariate(|y, x| {
+            if y == OuputCarry::Propagated as u64 {
+                x
+            } else {
+                y
+            }
+        });
+
+        let mut space = 1;
+        let mut step_output = carry_out.clone();
+        for _ in 0..=num_steps {
+            step_output[space..num_blocks]
+                .par_iter_mut()
+                .enumerate()
+                .for_each(|(i, block)| {
+                    let prev_block_carry = &carry_out[i];
+                    self.key.unchecked_apply_lookup_table_bivariate_assign(
+                        block,
+                        prev_block_carry,
+                        &lut_carry_propagation_sum,
+                    )
+                });
+            for i in space..num_blocks {
+                carry_out[i].copy_from(&step_output[i]);
+            }
+
+            space *= 2;
+        }
+
+        // The output carry of block i-1 becomes the input
+        // carry of block i
+        carry_out.rotate_right(1);
+        lhs.blocks
+            .par_iter_mut()
+            .zip(carry_out.par_iter())
+            .enumerate()
+            .for_each(|(i, (block, input_carry))| {
+                if i != 0 {
+                    self.key.unchecked_add_assign(block, input_carry);
+                }
+                self.key.message_extract_assign(block);
+            });
+    }
+
+    /// This add_assign two numbers
+    ///
+    /// It is after the Blelloch algorithm to do
+    /// prefix sum / cumulative sum in parallel.
+    ///
+    /// It is not "work efficient" as in, it does not adds
+    /// that much work compared to other parallel algorithm,
+    /// thus requiring less threads.
+    ///
+    /// However it is slower.
+    ///
+    /// At most num_block / 2 threads are used
+    ///
+    /// # Requirements
+    ///
+    /// - The parameters have 3 bits in total
+    /// - The input carries of both lhs and rhs must be empty
+    ///
+    /// # Output
+    ///
+    /// - lhs will have its carries empty
+    pub(crate) fn unchecked_add_assign_parallelized_work_efficient<PBSOrder: PBSOrderMarker>(
+        &self,
+        lhs: &mut RadixCiphertext<PBSOrder>,
+        rhs: &RadixCiphertext<PBSOrder>,
+        add_extra_one: AddExtraOne,
+    ) {
+        debug_assert!(lhs.block_carries_are_empty());
+        debug_assert!(rhs.block_carries_are_empty());
+        debug_assert!(self.key.message_modulus.0 * self.key.carry_modulus.0 >= (1 << 3));
+
+        let mut carry_out = self.add_and_generate_init_carry_array(lhs, rhs, add_extra_one);
+
+        let num_blocks = carry_out.len();
+        let num_steps = (carry_out.len() as f64).log2() as usize;
+
+        let lut_carry_propagation_sum = self.key.generate_accumulator_bivariate(|y, x| {
```

Comment:
the reversed y, x makes it harder to read, is there some more descriptive name that could be used here ? like

previous_block_carry, current_block_value or something ?

Reply:
changed that part to take (msb ,lsb) and added comment